src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,117 @@
+exist.
+
+Writer threads are then able to replace that object but cannot free it until the reader thread has
+left or released the generation. Instead the writer thread adds the object to a "stash" and it marks
+the stashed object with the generation indicating the point at which it was added to the stash. When
+no reader in the system holds a generation which is older than or equal to the stashed object's
+generation it can be freed.
+
+@section atomics Atomics
+
+As mentioned in @ref portability WiredTiger expects loads and stores of specific sizes to be atomic
+and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
+- Compare and Swap
+- Fetch add / sub
+- Add / sub fetch
+
+Currently WiredTiger is undergoing a transition where all shared memory variable access will be
+wrapped by an atomic function, this will enable TSan to run on the code base. To facilitate this
+transition the atomic load and store macros have been added to WiredTiger, these macros placate
+TSan. To prevent them from introducing a performance regression they utilize relaxed consistency.
+
+@section memory-barriers Memory Barriers
+
+WiredTiger utilizes memory barriers in a number of places to manage CPU instruction reordering and
+compiler reordering. Specifically CPUs are allowed to reorder load and store instructions. If this
+occurs when two threads are reading or writing to the same shared memory locations it can result in
+those threads seeing what would otherwise be incorrect state.
+
+WiredTiger defines barriers by describing which kind of reordering they prevent, for example a
+LoadLoad barrier prevents load instructions before the barrier from being reordered with load
+instructions after the barrier. Using this definition we can define 4 types of reorderings,
+LoadLoad, StoreStore, LoadStore and StoreLoad. To simplify the barrier semantics further
+WiredTiger defines acquire and release barriers.
+
+@subsection acq-bar Acquire barrier
+
+The acquire barrier prevents LoadLoad and LoadStore reorderings. It is accessible through two macros
+firstly the \c WT_ACQUIRE_BARRIER, this macro is just the barrier and nothing else. Most usages of
+the barrier are concerned with a single variable being reordered with many. This gives us the
+\c WT_ACQUIRE_READ_WITH_BARRIER which does a load from memory into a local variable before the
+barrier.
+
+@subsection rel-bar Release barrier
+
+The release barrier prevents StoreStore and LoadStore reorderings. Like the acquire barrier it is
+accessible through two macros. The \c WT_RELEASE_BARRIER, this like the matching acquire barrier
+macro is just the barrier. The \c WT_RELEASE_WRITE_WITH_BARRIER macro issues a barrier before a
+single store to memory. These barriers should be used as pairs. If one location requires an acquire
+barrier then there must be another location that requires a release barrier.
+
+Previously WiredTiger defined a \c WT_PUBLISH macro which prevented StoreStore reordering, this
```

Comment:
I feel we should further downplay this. May be we should remove this section.

Reply:
This discussion is tracked here: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612719663.